File: epi_judge_python/prime_sieve.py
from typing import List
import math

from test_framework import generic_test


# Given n, return all primes up to and including n.
def generate_primes(n: int) -> List[int]:
    # result = []
    # for i in range(2, n + 1):
    #     lim = math.floor(math.sqrt(i))
    #     prime = True
    #     for j in range(2, lim + 1):
    #         if i % j == 0:
    #             prime = False
    #     if prime:
    #         result.append(i)
    # return result
    primes = [False, False] + [True] * (n - 1)
    for i in range(n + 1):
        if primes[i]:
            m = 2
            while m * i <= n:
                primes[m * i] = False
                m += 1
    return [i for i in range(n + 1) if primes[i]]


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "prime_sieve.py", "prime_sieve.tsv", generate_primes
        )
    )

File: epi_judge_cpp/prime_sieve.cc
#include <algorithm>
#include <bitset>
#include <deque>
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
// Given n, return all primes up to and including n.
vector<int> GeneratePrimes(int n) {
  // TODO - you fill in here.
  // return {};

  if (n < 2) {
    return {};
  }
  std::deque<bool> sieve(n + 1, true);
  sieve[0] = sieve[1] = false;
  for (auto i = 0; i <= n; i++) {
    if (sieve[i]) {
      for (auto j = i + 1; j <= n; j++) {
        if (sieve[j] && (j % i == 0)) {
          sieve[j] = false;
        }
      }
    }
  }

  vector<int> result;
  for (auto i = 0; i <= n; i++) {
    if (sieve[i]) {
      result.emplace_back(i);
    }
  }
  return result;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"n"};
  return GenericTestMain(args, "prime_sieve.cc", "prime_sieve.tsv",
                         &GeneratePrimes, DefaultComparator{}, param_names);
}
